das/aggregator.go

```diff
@@ -58,49 +60,63 @@ func NewServiceDetails(service DataAvailabilityService, pubKey blsSignatures.Pub
 func NewAggregator(config AggregatorConfig, services []ServiceDetails) (*Aggregator, error) {
 	var aggSignersMask uint64
+	pubKeys := []blsSignatures.PublicKey{}
 	for _, d := range services {
 		if bits.OnesCount64(d.signersMask) != 1 {
 			return nil, fmt.Errorf("Tried to configure backend DAS %v with invalid signersMask %X", d.service, d.signersMask)
 		}
 		aggSignersMask |= d.signersMask
+		pubKeys = append(pubKeys, d.pubKey)
 	}
 	if bits.OnesCount64(aggSignersMask) != len(services) {
 		return nil, errors.New("At least two signers share a mask")
 	}
 
+	keyset := &arbstate.DataAvailabilityKeyset{
+		AssumedHonest: uint64(config.AssumedHonest),
+		PubKeys:       pubKeys,
+	}
+	ksBuf := bytes.NewBuffer([]byte{})
+	if err := keyset.Serialize(ksBuf); err != nil {
+		return nil, err
+	}
+	keysetHashBuf, err := keyset.Hash()
+	if err != nil {
+		return nil, err
+	}
+	var keysetHash [32]byte
+	copy(keysetHash[:], keysetHashBuf)
+
 	return &Aggregator{
 		config:                         config,
 		services:                       services,
 		requiredServicesForStore:       len(services) + 1 - config.AssumedHonest,
 		maxAllowedServiceStoreFailures: config.AssumedHonest - 1,
+		keysetHash:                     keysetHash,
+		keysetBytes:                    ksBuf.Bytes(),
 	}, nil
 }
 
 // Retrieve calls  on each backend DAS in parallel and returns immediately on the
 // first successful response where the data matches the requested hash. Otherwise
 // if all requests fail or if its context is canceled (eg via TimeoutWrapper) then
 // it returns an error.
-func (a *Aggregator) Retrieve(ctx context.Context, cert []byte) ([]byte, error) {
-	requestedCert, err := arbstate.DeserializeDASCertFrom(bytes.NewReader(cert))
-	if err != nil {
-		return nil, err
-	}
-
+func (a *Aggregator) Retrieve(ctx context.Context, cert *arbstate.DataAvailabilityCertificate) ([]byte, error) {
 	// Cert is the aggregate cert, validate it against DAS public keys
 	var servicesThatSignedCert []ServiceDetails
 	var pubKeys []blsSignatures.PublicKey
 	for _, d := range a.services {
-		if requestedCert.SignersMask&d.signersMask != 0 {
+		if cert.SignersMask&d.signersMask != 0 {
 			servicesThatSignedCert = append(servicesThatSignedCert, d)
 			pubKeys = append(pubKeys, d.pubKey)
 		}
 	}
 	if len(servicesThatSignedCert) < a.requiredServicesForStore {
-		return nil, fmt.Errorf("Cert %v was only signed by %d DASes, %d required.", requestedCert, len(servicesThatSignedCert), a.requiredServicesForStore)
+		return nil, fmt.Errorf("Cert %v was only signed by %d DASes, %d required.", cert, len(servicesThatSignedCert), a.requiredServicesForStore)
 	}
 
-	signedBlob := serializeSignableFields(*requestedCert)
-	sigMatch, err := blsSignatures.VerifySignature(requestedCert.Sig, signedBlob, blsSignatures.AggregatePublicKeys(pubKeys))
+	signedBlob := serializeSignableFields(cert)
+	sigMatch, err := blsSignatures.VerifySignature(cert.Sig, signedBlob, blsSignatures.AggregatePublicKeys(pubKeys))
```

Comment:
I think a lot of this is duplicated from `cert.VerifyNonPayloadParts`. Can we just call that here?

Reply:
In fact, do we even need these checks here, if `cert.VerifyNonPayloadParts` is already called before this in `parseSequencerMessage`?

Reply:
I don't understand this comment. This aggregator code is called from places other than `parseSequencerMessage`, so we need to check the property independently here.

Reply:
If we need the checks here, we should just call `cert.VerifyNonPayloadParts` here, unless there's a difference I missed between this verification and cert's